components/javascript-wrapper.js

```diff
@@ -0,0 +1,157 @@
+// can be sent through the output port to the next component
+var defaultUpdater = function() { 
+    return { out: arguments };
+}
+
+// Introspect the updater to determine what its parameters are.
+// The algorithm here comes from the following sources:
+//  - http://stackoverflow.com/questions/6921588/is-it-possible-to-reflect-the-arguments-of-a-javascript-function#answer-13660631
+//  - https://github.com/angular/angular.js/blob/master/src/auto/injector.js
+var introspect = function( fn ) {
+
+    var FN_ARGS = /^function\s*[^\(]*\(\s*([^\)]*)\)/m;
+    var FN_ARG_SPLIT = /,/;
+    var FN_ARG = /^\s*(_?)(\S+?)\1\s*$/;
+    var STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/mg;
+  
+    try {
+        var fnText = fn.toString().replace(STRIP_COMMENTS, '');
+        var argDecl = fnText.match(FN_ARGS);
+        var rawArgs = argDecl[1].split(FN_ARG_SPLIT);
+
+        var args = []; // default to no args
+        if ( ! _.isEqual(rawArgs, [''] )) { 
+            // Got some args so extract them
+            args = _.map( rawArgs, function( arg ) { 
+                return arg.replace(FN_ARG, function(all, underscore, name) { return name });
+            });
+        }
+
+        return args;
+    } catch (e) { 
+        // Could throw exception if we do not have source to introspect
+        // Use default of empty argument list
+        return [];
+    }
+};
+
+var ondata = function( portName, updaterArgs, nodeDef, payload ) {
+
+     // TODO: Integrate with James' work
+     if ( _.isUndefined( this.rpf ) ) { 
+         this.rpf = rpf;
+     } 
+
+     // First, save the data to the state
+     var vni = this.rpf.vni('');
+     vni.inputState( portName, 'test', '*', payload );
+
+     // Do we have data for our input ports?  
+     var allInputStates = vni.allInputStates();
+     var inPorts = _.isArray(nodeDef.inPorts) ? _.object(nodeDef.inPorts, []) : nodeDef.inPorts;
+     var required = _.compact(_.map(inPorts, function(port, name){
+         return port && port.required && name;
+     }));
+     var missing = _.difference(required, _.keys(allInputStates));
+     if ( ! _.isEmpty( missing ) ) {
+        // Don't have all port data yet - can't call updater so just return
+       return; 
+
+     } else { 
+        // Do we have any updaterArgs?  We might not if we do not have source code
+        if ( _.isEmpty( updaterArgs ) ) { 
+           // No updaterArgs, so just pass the state 
+           return nodeDef.updater.call( this, this.rpf.state );
+
+        } else {
+
+           // Have one or more parameters - get them in the right order and pass them 
+           var parameters = _.map( updaterArgs, function( arg ) { 
+               var parameter = vni.inputState( arg, 'test', '*' );
+               if ( _.isUndefined(parameter) && nodeDef.outPorts && nodeDef.outPorts.hasOwnProperty( arg ) ) {
+                   // passing in the name of the output port - is this correct? 
```

Comment:
If one of the parameters is called 'output', then vni.output.data should be passed as its value.